File: src/services/audit/detectors/ad/compliance/disa/audit-policies.ts
/**
 * DISA STIG V-220858 - Audit Policies
 * Checks DISA STIG audit policy requirements
 */

import { Finding } from '../../../../../../types/finding.types';
import { GpoSecuritySettings } from '../../../../../../providers/smb/smb.provider';

export function detectDisaAuditPolicies(
  gpoSettings: GpoSecuritySettings | null,
  _includeDetails: boolean
): Finding {
  const issues: string[] = [];

  if (!gpoSettings?.auditPolicies || gpoSettings.auditPolicies.length === 0) {
    issues.push('Audit policy not configured (V-220858)');
  } else {
    const audit = gpoSettings.auditPolicies;

    // DISA requires specific audit categories
    const disaRequirements = [
      { name: 'Account Logon (Success)', check: (a: typeof audit) => a.some((p) => p.category.includes('Logon') && p.success) },
      { name: 'Account Logon (Failure)', check: (a: typeof audit) => a.some((p) => p.category.includes('Logon') && p.failure) },
      { name: 'Account Management (Success)', check: (a: typeof audit) => a.some((p) => p.category.includes('Account Management') && p.success) },
      { name: 'Account Management (Failure)', check: (a: typeof audit) => a.some((p) => p.category.includes('Account Management') && p.failure) },
      { name: 'Policy Change (Success)', check: (a: typeof audit) => a.some((p) => p.category.includes('Policy Change') && p.success) },
      { name: 'Policy Change (Failure)', check: (a: typeof audit) => a.some((p) => p.category.includes('Policy Change') && p.failure) },
      { name: 'Privilege Use (Success)', check: (a: typeof audit) => a.some((p) => p.category.includes('Privilege Use') && p.success) },
      { name: 'Privilege Use (Failure)', check: (a: typeof audit) => a.some((p) => p.category.includes('Privilege Use') && p.failure) },
    ];

    for (const req of disaRequirements) {
      if (!req.check(audit)) {
        issues.push(`${req.name} audit not enabled (V-220858)`);
      }
    }
  }

  return {
    type: 'DISA_AUDIT_POLICIES',
    severity: 'high',
    category: 'compliance',
    title: 'DISA STIG - Audit Policies Non-Compliant',
    description:
      'Audit policies do not comply with DISA STIG V-220858. All required audit categories should be enabled.',
    count: issues.length > 0 ? 1 : 0,
    details: issues.length > 0 ? { violations: issues, framework: 'DISA', control: 'V-220858' } : undefined,
  };
}

File: internal/audit/detectors/ad/compliance/disa/audit-policies.go
package disa

import (
	"context"

	"github.com/etcsec-com/etc-collector/internal/audit"
	"github.com/etcsec-com/etc-collector/pkg/types"
)

// AuditPoliciesDetector checks DISA STIG audit policy compliance
type AuditPoliciesDetector struct {
	audit.BaseDetector
}

// NewAuditPoliciesDetector creates a new detector
func NewAuditPoliciesDetector() *AuditPoliciesDetector {
	return &AuditPoliciesDetector{
		BaseDetector: audit.NewBaseDetector("DISA_AUDIT_POLICIES", audit.CategoryCompliance),
	}
}

// Detect executes the detection
func (d *AuditPoliciesDetector) Detect(ctx context.Context, data *audit.DetectorData) []types.Finding {
	// DISA STIG audit policy requirements cannot be fully verified via LDAP
	// These settings are typically in GPOs and require additional tools to audit

	finding := types.Finding{
		Type:        d.ID(),
		Severity:    types.SeverityMedium,
		Category:    string(d.Category()),
		Title:       "DISA STIG Audit Policies Review Required",
		Description: "DISA STIG audit policy compliance cannot be verified via LDAP. Manual review of audit policies recommended.",
		Count:       0,
		Details: map[string]interface{}{
			"framework": "DISA",
			"stig":      "Windows Server STIG",
			"note":      "Audit policy settings require GPO analysis",
			"requiredAuditCategories": []string{
				"V-63447: Account Logon - Credential Validation",
				"V-63449: Account Management - Security Group Management",
				"V-63451: Account Management - User Account Management",
				"V-63453: Detailed Tracking - Process Creation",
				"V-63455: Logon/Logoff - Logon",
				"V-63457: Logon/Logoff - Special Logon",
				"V-63459: Object Access - Removable Storage",
				"V-63461: Policy Change - Audit Policy Change",
				"V-63463: Privilege Use - Sensitive Privilege Use",
				"V-63465: System - Security State Change",
				"V-63467: System - Security System Extension",
			},
		},
	}

	return []types.Finding{finding}
}

func init() {
	audit.MustRegister(NewAuditPoliciesDetector())
}
